packages/adblocker/src/engine/engine.ts

```diff
@@ -86,16 +86,72 @@ export interface Caching {
   write: (path: string, buffer: Uint8Array) => Promise<void>;
 }
 
-export default class FilterEngine extends EventEmitter<
-  | 'csp-injected'
-  | 'html-filtered'
-  | 'request-allowed'
-  | 'request-blocked'
-  | 'request-redirected'
-  | 'request-whitelisted'
-  | 'script-injected'
-  | 'style-injected'
-> {
+// Cosmetic rule is commonly used word in the project,
+// but cosmetic filter is used for better documentation.
+export type NetworkFilterMatchingContext = {
+  request: Request;
+
+  // `type` will cause a conflict in the variable name of `Request`.
+  matchType: FilterType.NETWORK;
+};
+
+// We do have a full context in case of the network filter matching.
+// It's because network filter matching does rely on an argument of the request.
+type CosmeticFilterMatchingContextBase = {
+  url: string;
+  domain: string | null | undefined;
+  hostname: string | undefined;
+
+  // Additional context given from user
+  matchType: FilterType.COSMETIC;
+  reference: any;
+};
+
+export type CosmeticFilterMatchingContext = CosmeticFilterMatchingContextBase &
+  Partial<{
+    classes: string[] | undefined;
+    hrefs: string[] | undefined;
+    ids: string[] | undefined;
+
+    allowGenericHides: boolean;
+    allowSpecificHides: boolean;
+
+    getBaseRules: boolean;
+    getInjectionRules: boolean;
+    getExtendedRules: boolean;
+    getRulesFromDOM: boolean;
+    getRulesFromHostname: boolean;
+  }>;
```

Comment:
Can we use some type helper function like `ReturnType` instead of copying the type?

Reply:
Mixed the type from `FilterEngine.getCosmeticFilters` and `CosmeticFilterBucket.getCosmeticFilters`.